client/webserver/websocket.go

```diff
@@ -0,0 +1,162 @@
+// and a websocketHandler thread.
+func (s *webServer) handleWS(w http.ResponseWriter, r *http.Request) {
+	// If the IP address includes a port, remove it.
+	ip := r.RemoteAddr
+	// If a host:port can be parsed, the IP is only the host portion.
+	host, _, err := net.SplitHostPort(ip)
+	if err == nil && host != "" {
+		ip = host
+	}
+	wsConn, err := ws.NewConnection(w, r, pingPeriod+pongWait)
+	if err != nil {
+		log.Errorf("ws connection error: %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	go s.websocketHandler(wsConn, ip)
+}
+
+// websocketHandler handles a new websocket client by creating a new wsClient,
+// starting it, and blocking until the connection closes. This method should be
+// run as a goroutine.
+func (s *webServer) websocketHandler(conn ws.Connection, ip string) {
+	log.Tracef("New websocket client %s", ip)
+	// Create a new websocket client to handle the new websocket connection
+	// and wait for it to shutdown.  Once it has shutdown (and hence
+	// disconnected), remove it.
+	var wsLink *ws.WSLink
+	wsLink = ws.NewWSLink(ip, conn, pingPeriod, func(msg *msgjson.Message) *msgjson.Error {
+		return s.handleMessage(wsLink, msg)
+	})
+	wsLink.Start()
+	wsLink.WaitForShutdown()
+	log.Tracef("Disconnected websocket client %s", ip)
+}
+
+// handleMessage handles the websocket message, calling the right handler for
+// the route.
+func (s *webServer) handleMessage(conn *ws.WSLink, msg *msgjson.Message) *msgjson.Error {
+	log.Tracef("message of type %d received for route %s", msg.Type, msg.Route)
+	if msg.Type == msgjson.Request {
+		handler, found := wsHandlers[msg.Route]
+		if !found {
+			return msgjson.NewError(msgjson.UnknownMessageType, "unknown route '"+msg.Route+"'")
+		}
+		return handler(s, conn, msg)
+	}
+	// Web server doesn't send requests, only responses and notifications, so
+	// a response-type message from a client is an error.
+	return msgjson.NewError(msgjson.UnknownMessageType, "web server only handles requests")
+}
+
+// wsHandlers is the map used by the server to locate the router handler for a
+// request.
+var wsHandlers = map[string]func(*webServer, *ws.WSLink, *msgjson.Message) *msgjson.Error{
+	"loadmarket": wsLoadMarket,
+	"unmarket":   wsUnmarket,
+}
+
+// marketLoad is sent by websocket clients to subscribe to a market and request
+// the order book.
+type marketLoad struct {
+	DEX    string `json:"dex"`
+	Market string `json:"market"`
+}
+
+// marketResponse is the websocket update sent when the client requests a
+// market via the 'loadmarket' route.
+type marketResponse struct {
+	Book         *core.OrderBook `json:"book"`
+	Market       string          `json:"market"`
+	DEX          string          `json:"dex"`
+	Base         string          `json:"base"`
+	Quote        string          `json:"quote"`
+	BaseBalance  float64         `json:"baseBalance"`
+	QuoteBalance float64         `json:"quoteBalance"`
```

Comment:
I don't trust JSON representations of float64 to do the trick reliably.  How about int64?  Well, maybe it's OK since this is the data going to the web ui, but perhaps `int64` from `clientCore.Balance()` since I expect other client subsystems to want a balance representation with no possibility for precision loss or ambiguity regarding atoms vs. other unit.

Reply:
Switched to `uint64`.